lib/galaxy/jobs/runners/__init__.py

```diff
@@ -678,14 +678,16 @@ def __init__(self, job_wrapper: "JobWrapper", job_destination: "JobDestination")
         self.job_wrapper = job_wrapper
         self.job_destination = job_destination
         self.runner_state = None
-        self.exit_code_file = default_exit_code_file(job_wrapper.working_directory, job_wrapper.get_id_tag())
-
         self.redact_email_in_job_name = True
         if self.job_wrapper:
             self.redact_email_in_job_name = self.job_wrapper.app.config.redact_email_in_job_name
 
         self.cleanup_file_attributes = ["job_file", "output_file", "error_file", "exit_code_file"]
 
+    @property
+    def exit_code_file(self) -> str:
+        return default_exit_code_file(self.job_wrapper.working_directory, self.job_wrapper.get_id_tag())
```

Comment:
This breaks runners like pbs that instantiate `AsynchronousJobState(..., exit_code_file="some path", ...)`, which would then fail with `AttributeError: can't set attribute 'exit_code_file'` .

Maybe check an internal variable `self._exit_code_file` (initiated as `None`) and if not set assign the calculated value to it before returning it. Modify `AsynchronousJobState.__init__()` to set `self._exit_code_file` instead of `self.exit_code_file` .